platforms/avr/drivers/ws2812_spi.c

```diff
@@ -0,0 +1,64 @@
+/*
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This is a SPI driver for the SK6812/WS2812 RGB leds,
+ * for the ATmega32U.
+ *
+ * It sets SPI to 8MHz and sends 11100000 for a zero,
+ * or 11111110 for a one. The extra time to prepare the
+ * next bit ensures the timming after each bit.
+ *
+ * It requires a 16MHz CPU clock and the pins PB0 and PB1
+ * unused (they are part of the ISP interface, so it's not
+ * uncommon), and the leds connected to the PB2 pin. Also,
+ * it can manage a maximum of 85 leds.
+ *
+ * It uses 40 bytes less than the WS2812 driver.
+ */
+
+#include <avr/interrupt.h>
+#include "rgblight.h"
+#include "ws2812.h"
+
+void ws2812_setleds(LED_TYPE *ledarray, uint8_t leds) {
+    // Initialize SPI for the WS2812/SK6812 driver
+    // Set PB2 and PB0 as output (not doing it freezes the SPI)
+    DDRB |= (1 << PB2);
+    DDRB |= (1 << PB0);
+    // Enable SPI, Master, set clock rate fck/2, MODE1
+    SPSR = 1 << SPI2X;
+    SPCR = (1 << SPE) | (1 << MSTR) | (1 << CPHA);
+    asm volatile("            mov  r24, %0\n\t"
+                 "            add  r24, r24\n\t"  // r24 = leds * 2
+                 "            add   %0, r24\n\t"  // %0 = leds * 3
+                 "loop_byte:  ld   r24, Z+\n\t"   // r24 contains the byte to send to the led
+                 "            ldi  r25, 8\n\t"    // 8 bits to send
+                 "loop_bits:  ldi  r26, 0xE0\n\t" // 11100000 -> zero bit (375ns up)
+                 "            sbrc r24, 7\n\t"
+                 "            ldi  r26, 0xF8\n\t" // 11111110 -> one bit (875ns up)
+                 "            out  0x2E, r26\n\t" // out SPDR, r26; begin the SPI transfer
+                 "            add  r24, r24\n\t"  // rotate data one bit to the left
+                 "wait_ready: in   r27, 0x2D\n\t" // in r27, SPSR
+                 "            andi r27, 0x80\n\t" // check SPIF to know if the transfer is complete.
+                 "            breq wait_ready\n\t"
+                 "            subi r25, 1\n\t" // are there more bits to send in this byte?
+                 "            brne loop_bits\n\t"
```

Comment:
So in the best case this code would take 9 CPU cycles between reading SPSR with the SPIF bit set and completing the next SPDR write; the delay would probably become at least 10 CPU cycles (and maybe more) before the SPI controller actually starts the send operation.  So the total bit time will be at least 10 + 2*8 = 26 CPU cycles, or 1625 ns at 16 MHz — this corresponds to about 600 kbit/s bit rate.  The usual WS2812 timings assume the bit time of 1250 ns (800 kbit/s bit rate).  So it looks like the SPI driver is slower than the existing bitbang driver — and that driver already has some speed problems with longer LED chains (with more than 33 RGB LEDs the send time goes above 1 ms when the bit time is 1250 ns; with 1625 ns that will happen if you go past 25 LEDs).

The increased bit time also probably violates the official datasheets of most WS2182-compatible LEDs; although in practice such violation would not matter unless the extra idle time is so large that it would exceed the reset threshold.

The driver also does not disable interrupts, unlike the existing bitbang driver; so apparently it expects that any interrupt handlers that may be invoked would complete faster than the LED reset threshold, which for some LEDs might be around 9 µs (or just 144 CPU cycles at 16 MHz): https://cpldcpu.wordpress.com/2014/01/14/light_ws2812-library-v2-0-part-i-understanding-the-ws2812/ (but apparently newer LEDs have much more relaxed reset timings, so you might not notice this problem in your testing).  In particular, the execution time of the USB interrupt handler would almost certainly exceed 144 CPU cycles in some cases, so the driver won't be reliable in this form; and disabling interrupts would make this driver not really better than the bitbang driver.

Reply:
I don't see why that time is a problem. Let's say that the LED chain is stable after an update. The microcontroller begins to send data; the first 24 bits are seen only by the first LED. The next 24 are seen by the first and second LEDs. The next 24 bits are seen by the first, second and third LEDs. And so on, so it doesn't matter that we need more than a millisecond, because during all that time all the LEDs that received a new color are seeing data and won't "reset". AFAIK, the color is updated only after 1us without "seeing" any data.

About the interrupts, you are right... fixing it.

Reply:
I did some little changes, and now it is as fast as the bit-banging function, because now it now prepares the next byte to send while the SPI interface is sending the previous one.